README.md
# Typy

[![Build Status](https://travis-ci.org/flexdinesh/typy.svg?branch=master)](https://travis-ci.org/flexdinesh/typy)
[![dependencies Status](https://david-dm.org/flexdinesh/typy/status.svg)](https://david-dm.org/flexdinesh/typy)
[![npm version](https://badge.fury.io/js/typy.svg)](https://www.npmjs.com/package/typy)
[![License: MIT](https://img.shields.io/badge/License-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![Gitter](https://badges.gitter.im/flexdinesh/typy.svg)](https://gitter.im/flexdinesh/typy?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

Type checking library for JavaScript with a _'sweeter'_ syntax.

`t('foo').isString // => true`

## New in version 3 🔥

- [Schema Validation](#isvalid-schema-validation)
- [Custom Types](#addcustomtypes-custom-types)

Version **3.0.0** introduces **BREAKING** changes (for node.js CommonJS style imports only).

```js
// Before v3.0.0, `t` function was imported as
const t = require('typy');

//From v3.0.0, `t` function should be imported as
const { t } = require('typy');

// Note: This version does not affect previous ES6 style imports._
import t, { Schema, addCustomTypes } from 'typy'; // this will still work
import { t, Schema, addCustomTypes } from 'typy'; // this will also work
```

## Why? [![start with why](https://img.shields.io/badge/start%20with-why%3F-brightgreen.svg?style=flat)](http://www.ted.com/talks/simon_sinek_how_great_leaders_inspire_action)

There are a hundred other type checking libraries out there. But **Typy** is built with three core behavioral aspects.

1. No surprises. **Typy** will never throw, no matter what the input is.
2. Object check will only look for **{ }** rather than JavaScript's native behavior of considering everything as objects such as arrays, functions, null, etc.
3. _Thought Driven Development_. Code should exactly mimic your thoughts on the logic rather than writing extra code just because that's how JavaScript works. `t(obj).isDefined // => true`
4. Custom type validation and schema validation.

## Install

```
$ npm install --save typy
```

## Usage

```js
import t from 'typy'; // ES6 style import
// var t = require('typy'); // CommonJS style import (version < 3)
// var t = require('typy').default; // CommonJS style import (version >= 3)

if (t('hello').isString) { // => true
  console.log('Input is a String!')
} else {
  console.log('Input is not a String!')
}

// More examples
t('22').isNumber // => false
t('22').isString // => true
t({}).isObject // => true
t([]).isArray // => true
t([]).isObject // => false

const sym = Symbol('typyIsAwesome');
t(sym).isSymbol // => true

// obj.goodKey.nestedKey = 'helloworld'
t(obj, 'goodKey.nestedKey').isDefined // => true
t(obj, 'badKey.nestedKey').isDefined // => false
// Typy won't throw undefined error for badKey.nestedKey

// to check if obj.goodKey.nestedKey is a string
t(obj, 'goodKey.nestedKey').isString // => true
t(obj, 'badKey.nestedKey').isString // => false

const deepObj = {
  nestedKey: {
    goodKey: 'hello',
    superNestedKey: {}
  }
};
// safely return the value from a nested key in an object
const myObj = t(deepObj, 'nestedKey.goodKey').safeObject; // => 'hello'
// Typy won't throw undefined error for badKey.goodKey
// instead the return value will be undefined
const myObj = t(deepObj, 'badKey.goodKey').safeObject; // => undefined
```

## API

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
  - [t(input, optionalObjectPath)](#tinput-optionalobjectpath)
  - [isDefined](#isdefined)
  - [isUndefined](#isundefined)
  - [isNull](#isnull)
  - [isNullOrUndefined](#isnullorundefined)
  - [isBoolean](#isboolean)
  - [isTrue](#istrue)
  - [isFalse](#isfalse)
  - [isTruthy](#istruthy)
  - [isFalsy](#isfalsy)
  - [isObject](#isobject)
  - [isEmptyObject](#isemptyobject)
  - [isString](#isstring)
  - [isEmptyString](#isemptystring)
  - [isNumber](#isnumber)
  - [isArray](#isarray)
  - [isEmptyArray](#isemptyarray)
  - [isFunction](#isfunction)
  - [isDate](#isdate)
  - [isSymbol](#issymbol)
  - [safeObject](#safeobject)
  - [safeObjectOrEmpty](#safeobjectorempty)
  - [safeString](#safestring)
  - [safeNumber](#safenumber)
  - [safeBoolean](#safeboolean)
  - [safeFunction](#safefunction)
  - [safeArray](#safearray)
  - [isValid (Schema Validation)](#isvalid-schema-validation)
  - [addCustomTypes (Custom Types)](#addcustomtypes-custom-types)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

#### t(input, optionalObjectPath)

Pass in your input to the t() method and **Typy** will take care of everything

```js
// you can pass any type of input
// Number, String, Object, null, undefined, Array, anything
t('str')
t(22)
t({foo: 'fooooo', bar: 'barooo'})
t([2, 'three', 'hey'])

const obj = {
  goodKey: {
    nestedKey: 'hello world'
  }
}
// To pass nested path of an object
// Ex. obj.goodKey.nestedKey
// You have to pass the path as string in the second param
t(obj, 'goodKey.nestedKey')
t(obj, 'badKey.nestedKey')
// this is because if you pass t(obj.badKey.nestedKey),
// you will get undefined exception
// because that is how javascript is designed
// to overcome that we need to pass the sub key as a string to Typy
```

#### isDefined

Returns _true_ if the input is defined.

```js
const obj = {
  goodKey: 'hello'
}

t(obj.goodKey).isDefined // => true
t(obj.badKey).isDefined // => false
```

#### isUndefined

Returns _true_ if the input is undefined.

```js
const obj = {
  goodKey: 'hello'
}

t(obj.goodKey).isUndefined // => false
t(obj.badKey).isUndefined // => true
```

#### isNull

Returns _true_ if the input is null.

```js
const obj = {
  foo: null
}

t(obj.foo).isNull // => true
```

#### isNullOrUndefined

Returns _true_ if the input is null or undefined.

```js
const obj = {
  foo: null
}

t(obj.foo).isNullOrUndefined // => true
t(obj.bar).isNullOrUndefined // => true
```

#### isBoolean

Returns _true_ if the input is either `true` or `false`.

```js
t(true).isBoolean // => true
t(false).isBoolean // => true
```

#### isTrue

Returns _true_ if the input is Boolean `true`.

```js
t(true).isTrue // => true
t(false).isTrue // => false
```

#### isFalse

Returns _true_ if the input is Boolean `false`.

```js
t(true).isFalse // => false
t(false).isFalse // => true
```

#### isTruthy

Returns _true_ if the input is considered _truthy_.

In JavaScript anything other than `false`, `0`, `''`, `""`, `null`, `undefined` and `NaN` is considered _truthy_.

```js
t('Typy is amazing =)').isTruthy // => true
t({}).isTruthy // => true
t(22).isTruthy // => true
t([1, 'two']).isTruthy // => true
```

#### isFalsy

Returns _true_ if the input is considered _falsy_.

In JavaScript any of these values `false`, `0`, `''`, `""`, `null`, `undefined` and `NaN` are considered _falsy_.

```js
t(0).isFalsy // => true
t(null).isFalsy // => true
t(undefined).isFalsy // => true
t(false).isFalsy // => true
```


#### isObject

Returns _true_ if the input is an object.

```js
const obj = {
  foo: null
}

t(obj).isObject // => true
t({}).isObject // => true
```

_Note: Only { } objects will return this as true as opposed to javascript definition of Object which includes Arrays, Functions, anything and everything related to prototype. This is an intentional behavior as we don't want arrays to return true for isObject._

#### isEmptyObject

Returns _true_ if the input is an empty object, _aka_ object without any keys.

```js
const obj = {
  foo: 'hello there',
  bar: {}
}

t(obj.bar).isEmptyObject // => true
t({}).isEmptyObject // => true
t(obj).isEmptyObject // => false
```

#### isString

Returns _true_ if the input is a string.

```js
const obj = {
  foo: 'typy is awesome =)',
}
t(obj.foo).isString // => true
t('').isString // => true
t(22).isString // => false
t(null).isString // => false
```

#### isEmptyString

Returns _true_ if the input is an empty string.

```js
t('').isEmptyString // => true
t('typy is so great').isEmptyString // => false
```

#### isNumber

Returns _true_ if the input is a number.

```js
t(22).isNumber // => true
t('i am a string').isNumber // => false
t({}).isNumber // => false
```

#### isArray

Returns _true_ if the input is an array.

```js
t([]).isArray // => true
t([1, 2, 'typy']).isArray // => true
t({}).isArray // => false
```

#### isEmptyArray

Returns _true_ if the input is an empty array.

```js
t([]).isEmptyArray // => true
t([1, 2, 'typy']).isEmptyArray // => false
```

#### isFunction

Returns _true_ if the input is a function.

```js
const func = () => {};
t(func).isFunction // => true
t({}).isFunction // => false
```

#### isDate

Returns _true_ if the input is a javascript's date object.

```js
const date = new Date();
t(date).isDate // => true
t({}).isDate // => false
```

#### isSymbol

Returns _true_ if the input is a javascript's Symbol.

```js
const mySym = Symbol(123);
const anotherSymbol = Symbol('typyIsAwesome');

t(mySym).isSymbol // => true;
t(Object(anotherSymbol)).isSymbol  // => true;

t({}).isSymbol // => false
t([]).isSymbol // => false
t(null).isSymbol // => false
```

#### safeObject

Safely returns the value from a nested object path without throwing any error.

```js
const deepObj = {
  nestedKey: {
    goodKey: 'hello',
    superNestedKey: {}
  }
};
// Typy can safely return the value from a nested key in an object
const myObj = t(deepObj, 'nestedKey.goodKey').safeObject; // => 'hello'
// Typy won't throw if the key at any level is not found
// instead will return undefined
const myObj = t(deepObj, 'badKey.goodKey').safeObject; // => undefined

const anotherDeepObj = {
  nestedArray: [{
    goodKey: 'hello one',
    superNestedKey: {}
  }, {
    goodKey: 'hello two',
    superNestedKey: {
      superGoodKey: 'typy is great :)'
    }
  }]
};
// Typy can safely return the value even from a nested key in a nested array
const myObj = t(anotherDeepObj, 'nestedArray[1].superNestedKey.superGoodKey').safeObject; // => 'typy is great :)'
```

#### safeObjectOrEmpty

Safely returns the value from a nested object path if the path exists
or returns an empty object if the.

```js
const deepObj = {
  nestedKey: {
    goodKey: 'hello',
    superNestedKey: {}
  }
};
// Typy can safely return the value from a nested key in an object
const myObj = t(deepObj, 'nestedKey.goodKey').safeObjectOrEmpty; // => 'hello'
// Typy won't throw if the key at any level is not found
// instead will return an empty object
const myObj = t(deepObj, 'badKey.goodKey').safeObjectOrEmpty; // => {}

const anotherDeepObj = {
  nestedArray: [{
    goodKey: 'hello one',
    superNestedKey: {}
  }, {
    goodKey: 'hello two',
    superNestedKey: {
      superGoodKey: 'typy is great :)'
    }
  }]
};
// Typy can safely return the value even from a nested key in a nested array
const myObj = t(anotherDeepObj, 'nestedArray[1].superNestedKey.superGoodKey').safeObjectOrEmpty; // => 'typy is great :)'
```

#### safeString

Returns the string value if the input type is string or will return an empty string `''`.

```js
const str = t('typy is safe').safeString; // => 'typy is safe'
const str = t(null).safeString; // => ''
const str = t(undefined).safeString; // => ''
const str = t(22).safeString; // => ''
```

#### safeNumber

Returns the number if the input type is Number or will return `0`.

```js
const num = t(22).safeNumber; // => 22
const num = t('22').safeNumber; // => 0
const num = t(undefined).safeNumber; // => 0
const num = t(null).safeNumber; // => 0
```

#### safeBoolean

Returns the boolean if the input type is Boolean or will return `false`.

```js
const bool = t(true).safeBoolean; // => true
const bool = t(false).safeBoolean; // => false
const bool = t('22').safeBoolean; // => false
const bool = t(undefined).safeBoolean; // => false
const bool = t(22).safeBoolean; // => false
```

#### safeFunction

Returns the function if the input type is function or will return an empty function `() => {}`.

```js
const helloFunc = () => { return 'Hello World!' }
const func = t(helloFunc).safeFunction; // => helloFunc reference
const func = t('I am a string').safeFunction; // => empty function () => {}
const func = t(undefined).safeFunction; // => empty function () => {}
const func = t(null).safeFunction; // => empty function () => {}
```

#### safeArray

Safely returns the value from a nested object path or an empty array. If the path specified exists but is not an array, returns an array containing the value of the specified path.

```js
const deepObj = {
  nestedKey: [
    {
      goodKey: ['hello'],
      numberKey: 10,
      superNestedKey: {}
    },
  ]
};
// Typy can safely return the value from a nested key in an object or an array
const myObj = t(deepObj, 'nestedKey').safeArray; // => [ { goodKey: ['hello'], numberKey: 10, superNestedKey: {} } ]
const myObj = t(deepObj, 'nestedKey[0].goodKey').safeArray; // => ['hello']
// Typy can wrap a value or object inside an array
const myObj = t(deepObj, 'nestedKey[0].numberKey').safeArray; // => [ 10 ]
const myObj = t(deepObj, 'nestedKey[0].superNestedKey').safeArray; // => [ {} ]
// Typy won't throw if the key at any level is not found
// instead will return an empty array
const myObj = t(deepObj, 'nestedKey[1]').safeArray; // => []
const myObj = t(deepObj, 'badKey.goodKey').safeArray; // => []
```

#### isValid (Schema Validation)

`isValid` is used to check and validate the schema of an object. It returns `true` if the schema of the object matches the schema passed or `false` if the schema doesn't match.

```js
import t, { Schema } from 'typy';

const superheroSchema = {
  name: Schema.String,
  age: Schema.Number,
  appearances: [
    {
      title: Schema.String,
      alias: Schema.String,
    }
  ],
  lastSeen: Schema.Date
};
const batmanObject = {
  name: 'Batman',
  age: 45,
  isAlive: true,
  appearances: [
    {
      title: 'The Dark Knight',
      alias: 'Bruce',
    }
  ],
  lastSeen: new Date(14894561568)
};
const isSchemaValid = t(batmanObject, superheroSchema).isValid; // true

const simpleSchema = {
  name: Schema.String,
  arr: Schema.Array
};
const obj = {
  name: 'Jack',
  arr: [1, 2, 3]
};
const isSchemaValid = t(obj, simpleSchema).isValid; // true
```

The following **Schema types** are available in typy.

- Number
- String
- Array
- Boolean
- Null
- Undefined
- Function
- Date

#### addCustomTypes (Custom Types)

`addCustomTypes` is used to pass custom validators to **Typy**. It can be used to validate any ipnut for custom types, like this `t(input).isMyCustomType`.

You will have to add custom types only once in the project (preferabby in entry file. ex. `index.js`)

Entry file (Ex. `index.js`)

```js
import t, { addCustomTypes } from 'typy';

addCustomTypes({
  isPhone: (input) => (t(input).isNumber && /^\d{10}$/g.test(String(input))), // has 10 digits
  isAddress: (input) => (t(input).isString && input.toUpperCase().includes('STREET')) // includes 'street' in input
});

```

Anywhere in the project

```js
import t from 'typy';

const isThePhoneNumberValid = t(9892389239).isPhone; // => true
const isThePhoneNumberValid = t('abcdefg').isPhone; // => false

const isTheAddressValid = t('10 Downing Street').isAddress; // => true
const isTheAddressValid = t('I like cats 🐈').isAddress; // => false

```

## Contributors

Thanks goes to these amazing people 🎉

| [<img src="https://avatars3.githubusercontent.com/u/5777880?v=4" width="100px;"/><br /><sub><b>Dinesh Pandiyan</b></sub>](https://github.com/flexdinesh)<br /> | [<img src="https://avatars0.githubusercontent.com/u/6686039?v=4" width="100px;"/><br /><sub><b>Ozair Patel</b></sub>](https://github.com/OzairP)<br /> | [<img src="https://avatars0.githubusercontent.com/u/23170622?v=4" width="100px;"/><br /><sub><b>Aneerudh</b></sub>](https://github.com/Aneedroid)<br /> | [<img src="https://avatars1.githubusercontent.com/u/13482258?v=4" width="100px;"/><br /><sub><b>Ruphaa Ganesh</b></sub>](https://github.com/ruphaa)<br /> | [<img src="https://avatars0.githubusercontent.com/u/4320434?v=4" width="100px;"/><br /><sub><b>Quentin Jadeau</b></sub>](https://github.com/jadok)<br /> | [<img src="https://avatars3.githubusercontent.com/u/416559?s=400&v=4" width="100px;"/><br /><sub><b>dan</b></sub>](https://github.com/danthewolfe)<br />
| :---: | :---: | :---: | :---: | :---: | :---: |
| [<img src="https://avatars2.githubusercontent.com/u/13327?v=4" width="100px;"/><br /><sub><b>Robert Schadek</b></sub>](https://github.com/burner)<br /> | [<img src="https://avatars0.githubusercontent.com/u/43963628?v=4" width="100px;"/><br /><sub><b>Michael Kirkpatrick</b></sub>](https://github.com/mlkirkpatrick)<br /> | [<img src="https://avatars1.githubusercontent.com/u/13729562?v=4" width="100px;"/><br /><sub><b>Ana Liza Pandac</b></sub>](https://github.com/analizapandac)<br /> | [<img src="https://avatars1.githubusercontent.com/u/49582824?v=4" width="100px;"/><br /><sub><b>Abdul Rehman</b></sub>](https://github.com/rehman-00001)<br /> | [<img src="https://avatars3.githubusercontent.com/u/13280256?v=4" width="100px;"/><br /><sub><b>laevilgenius</b></sub>](https://github.com/laevilgenius)<br /> 

## License

MIT © Dinesh Pandiyan


CHANGELOG.md
# Changelog

## To Be Released

### version (undecided)

* what's next

## Released

### 3.3.0 (17 Sep 2019)

* NEW: `t(dateObj).isDate` - is available.
* NEW: `t(Symbol('desc')).isSymbol` - is available
* FIX: `getNestedObject` breaks in ESM modules

### 3.2.2 (6 Sep 2019)

* TypeScript `index.d.ts` type changes

### 3.2.1 (29 Aug 2019)

* FIX: Whitelist TypeScript `index.d.ts` file for npm

### 3.2.0 (29 Aug 2019)

* NEW: TypeScript type definitions

### 3.0.1 (19 Jan 2019)

* Added support for bracket syntax while accessing nested objects

`t('someObject.firstArray[0].secondArray[0].complexObject["@id"]').isDefined` is now possible.

### 3.0.0 (9 Dec 2018)

* Schema Validation
  * **Typy** can validate if an object matches a specific type schema
* Custom Types
  * Custom type validators can be added to **Typy**

This version introduces **BREAKING** changes (for node.js imports only). Make sure you update your code accordingly.

Before v3.0.0, `t` function was imported as

```js
const t = require('typy');
```

From v3.0.0, `t` function should be imported as

```js
const { t } = require('typy');
```

_Note: Breaking changes apply only for node.js CommonJS style require statements. This version affect ES6 style imports._

### 2.0.1 (7 Mar 2018)

* Fix: Nested arrays with length > 10 will work like a charm now

### 2.0.0 (4 Mar 2018)

* Package minified with >60% compression
* New getters added
  * `safeNumber` returns the value if input is Number or returns 0
  * `safeBoolean` returns the value if input is Boolean or returns false
  * `safeFunction` returns the function if input is function or returns an empty function () => {}

### 1.4.3 (6 Feb 2018)

* Added support for use in TypeScript codebases

### 1.4.0 (6 Feb 2018)

* Two new getters are available in Typy
  * `safeObject` returns the value from a nested object path without throwing
  * `safeString` returns the value if input is string or returns empty string

### 1.3.0 (5 Feb 2018)

* Performance optimizations

### 1.2.0 (5 Feb 2018)

* Typy can now store results to a variable and check the type later

### 1.1.0 (4 Feb 2018)

* Fixed a bug for Number type validation

### 1.0.0 (4 Feb 2018)

* Typy is released to the world =)
* Following APIs are included in the first release
 * `isDefined` returns if the input is defined
 * `isUndefined` returns if the input is undefined
 * `isNull` returns if the input is null
 * `isNullOrUndefined` returns if the input is either null or undefined
 * `isBoolean` returns if the input is a boolean
 * `isTrue` returns if the input is boolean true
 * `isFalse` returns if the input is boolean false
 * `isTruthy` returns if the input is truthy
 * `isFalsy` returns if the input is falsy
 * `isFalsy` returns if the input is falsy
 * `isObject` returns if the input is an object - {}
 * `isEmptyObject` returns if the input is an empty object
 * `isString` returns if the input is a string
 * `isEmptyString` returns if the input is an empty string
 * `isNumber` returns if the input is a number
 * `isArray` returns if the input is an array
 * `isEmptyArray` returns if the input is an empty array
 * `isFunction` returns if the input is a function


test/util.js
import { assert } from 'chai';
import { getNestedObject, convertSchemaAndGetMatch, buildSchema, getSchemaMatch } from '../src/util';
import { Schema } from '../src/index';

describe('Nested Object/Keys Check', () => {
  it('should return nested object if exists', () => {
    const mockObj = { goodKey: 'hello' };
    assert(getNestedObject(mockObj, 'goodKey') !== undefined);
    const mockDeepObj = {
      nestedKey: {
        goodKey: 'hello'
      }
    };
    assert(getNestedObject(mockDeepObj, 'nestedKey.goodKey') === mockDeepObj.nestedKey.goodKey);
    assert(getNestedObject(mockDeepObj, 'nestedKey.goodKey') !== undefined);
  });

  it('should return undefined if nested object does not exist', () => {
    const mockObj = { goodKey: 'hello' };
    assert(getNestedObject(mockObj, 'goodKey.badkey') === undefined);
    const mockDeepObj = {
      nestedKey: {
        goodKey: 'hello'
      }
    };
    assert(getNestedObject(mockDeepObj, 'badKey.goodKey') === undefined);
    assert(getNestedObject(mockDeepObj, 'goodKey.badKey') === undefined);
    assert(getNestedObject(mockDeepObj, 'nestedKey.badKey') === undefined);
  });

  it('should return nested object when array element is in path', () => {
    let mockObj = {
      nestedArray: [
        { goodKey: 'hello one' },
        { goodKey: 'hello two' },
        { goodKey: 'hello three' },
        { goodKey: 'hello four' },
        { goodKey: 'hello five' },
        { goodKey: 'hello six' },
        { goodKey: 'hello seven' },
        { goodKey: 'hello eight' },
        { goodKey: 'hello nine' },
        { goodKey: 'hello ten' },
        { goodKey: 'hello eleven' },
      ]
    };
    assert(getNestedObject(mockObj, 'nestedArray[0].goodKey') === 'hello one');
    assert(getNestedObject(mockObj, 'nestedArray[1].goodKey') === 'hello two');
    assert(getNestedObject(mockObj, 'nestedArray[10].goodKey') === 'hello eleven');

    mockObj = {
      nestedArray: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k']
    };
    assert(getNestedObject(mockObj, 'nestedArray[0]') === 'a');
    assert(getNestedObject(mockObj, 'nestedArray[1]') === 'b');
    assert(getNestedObject(mockObj, 'nestedArray[10]') === 'k');
  });

  it('should return deeply nested object when there are multiple array elements in path', () => {
    const mockObj = {
      '@id': 'weird',
      firstArray: [
        {
          '@id': 'identifier',
          goodKey: 'hello one',
          secondArray: [
            {
              '@id': 'unique',
              goodKey: 'deep hello'
            }
          ],
          stackedArray: [['a1', 'a2'], ['b1', 'b2', 'b3'], ['c']]
        }
      ]
    };

    assert(getNestedObject(mockObj, '["@id"]') === 'weird');
    assert(getNestedObject(mockObj, 'firstArray[0].goodKey') === 'hello one');
    assert(getNestedObject(mockObj, 'firstArray[0]["@id"]') === 'identifier');
    assert(getNestedObject(mockObj, 'firstArray[0].secondArray[0].goodKey') === 'deep hello');
    assert(getNestedObject(mockObj, 'firstArray[0].secondArray[0][@id]') === 'unique');
    assert(getNestedObject(mockObj, 'firstArray[0].stackedArray[0][1]') === 'a2');
  });

  it('should return undefined when array element in path does not exist', () => {
    const mockObj = {
      nestedArray: [
        { goodKey: 'hello one' },
        { goodKey: 'hello two' }
      ]
    };
    assert(getNestedObject(mockObj, 'badArray[0].goodKey') === undefined);
    assert(getNestedObject(mockObj, 'nestedArray[1].goodKey.badkey') === undefined);
    assert(getNestedObject(mockObj, 'nestedArray[2].goodKey') === undefined);
    assert(getNestedObject(mockObj, 'nestedArray.goodKey') === undefined);
  });

  it('monkey test nested object', () => {
    const mockObj = { goodKey: 'hello' };
    assert(getNestedObject(mockObj, 'goodKey,badkey') === undefined);
    assert(getNestedObject(mockObj, null) === undefined);
    assert(getNestedObject(mockObj, 1) === undefined);
    assert(getNestedObject(mockObj, 0) === undefined);
    assert(getNestedObject(mockObj, {}) === undefined);
    assert(getNestedObject(undefined, 'goodKey,badkey') === undefined);
  });
});

describe('Objects and Schema check', () => {
  it('should return only typeOf for primitive types', () => {
    const objSchema = Schema.String;

    const mockSchemaObj = 'typy';

    assert(typeof buildSchema(objSchema) === typeof mockSchemaObj);
  });

  it('should convert complex schema to object', () => {
    const objSchema = {
      name: Schema.String,
      data: [
        {
          unitCount: Schema.Number,
          units: [
            {
              label: Schema.String,
              weight: Schema.Number,
            }
          ]
        }
      ]
    };

    const mockSchemaObj = {
      name: 'typy',
      data: [
        {
          unitCount: 1,
          units: [
            {
              label: 'typy',
              weight: 1,
            }
          ]
        }
      ]
    };

    const schema = buildSchema(objSchema);
    assert(schema.name === mockSchemaObj.name);
    assert(schema.data[0].unitCount === mockSchemaObj.data[0].unitCount);
    assert(schema.data[0].units[0].label === mockSchemaObj.data[0].units[0].label);
    assert(schema.data[0].units[0].weight === mockSchemaObj.data[0].units[0].weight);
  });

  it('should convert schema with function type to object with function', () => {
    const objSchema = {
      getName: Schema.Function,
      getRandomString: Schema.Function
    };

    const mockSchemaObj = {
      getName: () => {},
      getRandomString: function lol() { return true; } // eslint-disable-line object-shorthand
    };

    assert(typeof buildSchema(objSchema).getName === typeof mockSchemaObj.getName);
    assert(typeof buildSchema(objSchema).getRandomString === typeof mockSchemaObj.getRandomString);
  });

  it('should return true when two objects match', () => {
    const objectOne = {
      Rank: 1,
      Name: 'Popoye'
    };

    const objectTwo = {
      Rank: 99,
      Name: 'Wolverine'
    };

    assert(getSchemaMatch(objectOne, objectTwo) === true);
  });

  it('should return true for complex objects when they match', () => {
    const objectZero = {
      name: 'Batman',
      data: [
        {
          unitCount: 'luck',
          units: [
            {
              label: 'Superhero',
              weight: 85,
            }
          ]
        }
      ]
    };

    const objectOne = {
      name: 'Batman',
      data: [
        {
          unitCount: 13,
          units: [
            {
              label: 'Superhero',
              weight: 85,
            }
          ]
        }
      ]
    };

    const objectTwo = {
      name: 'typy',
      data: [
        {
          unitCount: 1,
          units: [
            {
              label: 'typy',
              weight: 1,
            }
          ]
        }
      ]
    };

    assert(getSchemaMatch(objectZero, objectTwo) === false);
    assert(getSchemaMatch(objectOne, objectTwo) === true);
  });

  it('should convert schema to object and return true for schema match', () => {
    const obj = {
      name: '1',
      data: [
        {
          unitCount: 10,
          units: [
            {
              label: 'lobel',
              weight: 10
            }
          ]
        }
      ]
    };

    const objSchema = {
      name: Schema.String,
      data: [
        {
          unitCount: Schema.Number,
          units: [
            {
              label: Schema.String,
              weight: Schema.Number,
            }
          ]
        }
      ]
    };

    const weirdObjSchema = {
      name: Schema.Number,
      data: [
        {
          unitCount: Schema.Number,
          units: [
            {
              label: Schema.String,
              weight: Schema.Number,
            }
          ]
        }
      ]
    };

    assert(convertSchemaAndGetMatch(obj, objSchema) !== -1);
    assert(typeof convertSchemaAndGetMatch(obj, objSchema) === 'object');
    assert(convertSchemaAndGetMatch(obj, weirdObjSchema) === -1);
  });
});


index.d.ts
declare module "typy" {

	type SafeFunction = (...args: any[]) => any;

	export function t (obj: any, nestedKeys?: string): Typy;
	export function addCustomTypes(validators: {[key: string]: any}): any;
  export default t;

  interface SchemaType {
   Number: number;
   String: string;
   Boolean: boolean;
   Null: null;
   Undefined: undefined;
   Array: any[];
   Function: () => void;
  }

  export const Schema: SchemaType;

	class Typy {

		t: (obj: any, nestedKeys?: string) => Typy

		readonly isValid: boolean
		readonly isDefined: boolean
		readonly isUndefined: boolean
		readonly isNull: boolean
		readonly isNullOrUndefined: boolean
		readonly isBoolean: boolean
		readonly isTrue: boolean
		readonly isFalse: boolean
		readonly isTruthy: boolean
		readonly isFalsy: boolean
		readonly isObject: boolean
		readonly isEmptyObject: boolean
		readonly isString: boolean
		readonly isEmptyString: boolean
		readonly isNumber: boolean
		readonly isArray: boolean
		readonly isEmptyArray: boolean
    readonly isFunction: boolean
		readonly isDate: boolean
		readonly isSymbol: boolean
		readonly safeObject: any
		readonly safeObjectOrEmpty: any
		readonly safeString: string
		readonly safeNumber: number
		readonly safeBoolean: boolean
		readonly safeFunction: SafeFunction
		readonly safeArray: any[]

	}

}


examples/example.js
/* eslint-disable no-console */
const { t, Schema, addCustomTypes } = require('../lib');

// Undefined Checks
const exampleObj = {
  goodKey: 'I exist :)',
  nestedKey: {
    nestedGoodKey: 'Me too!'
  },
  symbolKey: Symbol(123)
};

console.log(`exampleObj is defined - ${t(exampleObj).isDefined}`); // true
console.log(`exampleObj.goodKey is defined - ${t(exampleObj, 'goodKey').isDefined}`); // true
console.log(`exampleObj.badKey is defined - ${t(exampleObj, 'badKey').isDefined}`); // false
console.log(`exampleObj.badKey is undefined - ${t(exampleObj, 'badKey').isUndefined}`); // false
console.log(`exampleObj.nestedKey.nestedGoodKey is defined - ${t(exampleObj, 'nestedKey.nestedGoodKey').isDefined}`); // true
console.log(`exampleObj.badKey.nestedGoodKey is defined - ${t(exampleObj, 'badKey.nestedGoodKey').isDefined}`); // false
console.log(`exampleObj.badKey.nestedGoodKey is undefined - ${t(exampleObj, 'badKey.nestedGoodKey').isUndefined}`); // true

// Type Checks
const nullObj = null;
console.log(`nullObj is null - ${t(nullObj).isNull}`); // true

const str = 'hello';
console.log(`str is String - ${t(str).isString}`); // true

const num = 22;
console.log(`num is Number - ${t(num).isNumber}`); // true

let bool = true;
console.log(`bool is Boolean - ${t(bool).isBoolean}`); // true
bool = true;
console.log(`bool is true - ${t(bool).isTrue}`); // true
bool = false;
console.log(`bool is false - ${t(bool).isFalse}`); // true

const obj = {};
console.log(`obj is an Object - ${t(obj).isObject}`); // true
console.log(`obj is an empty Object - ${t(obj).isEmptyObject}`); // true

const arr = [];
console.log(`arr is an Array - ${t(arr).isArray}`); // true
console.log(`arr is an empty Array - ${t(arr).isEmptyArray}`); // true

const date = new Date();
console.log(`type of 'date' is Date - ${t(date).isDate}`); // true
console.log(`type of 'obj' is not date - ${t(obj).isDate}`); // false
console.log(`type of 'arr' is not date - ${t(arr).isDate}`); // false

const exampleDateObj = {
  key: {
    nestedGoodDate: 'new Date()'
  }
};
console.log(`exampleDateObj.key.nestedGoodDate is defined - ${t(exampleDateObj, 'key.nestedGoodDate').isDefined}`); // true
console.log(`exampleDateObj.key.nestedGoodDate is of type Date - ${t(exampleDateObj, 'key.nestedGoodDate').isDate}`); // true

const mySymbol = Symbol('symbol-description');
console.log(`is 'mySymbol' a Symbol - ${t(mySymbol).isSymbol}`);
console.log(`is 'obj' a Symbol? - ${t(obj).isSymbol}`);
console.log(`is 'arr' a Symbol? - ${t(arr).isSymbol}`);

// Schema Checks
const validSchema = {
  goodKey: Schema.String,
  nestedKey: {
    nestedGoodKey: Schema.String
  },
  symbolKey: Schema.Symbol
};
console.log(`exampleObj matches validSchema - ${t(exampleObj, validSchema).isValid}`); // true

const inValidSchema = {
  goodKey: Schema.String,
  nestedKey: Schema.Array
};
console.log(`exampleObj does not match inValidSchema - ${t(exampleObj, inValidSchema).isValid}`); // false

// Custom Types
addCustomTypes({
  isPhone: input => (t(input).isNumber && /^\d{10}$/g.test(String(input))), // has 10 digits
  isAddress: input => (t(input).isString && input.toUpperCase().includes('STREET')) // includes 'street' in address
});

const validPhoneNum = 9892389239;
const invalidPhoneNum = 98923892390;
console.log(`9892389239 is a valid phone num - ${t(validPhoneNum).isPhone}`); // true
console.log(`98923892390 is a valid phone num - ${t(invalidPhoneNum).isPhone}`); // false

const validAddress = '10 Downing Street';
const invalidAddress = 'bwahahaha';
console.log(`"10 Downing Street" is a valid address - ${t(validAddress).isAddress}`); // true
console.log(`"bwahahaha" is a valid address - ${t(invalidAddress).isAddress}`); // false
